File: Problems/Algorithms/126. Word Ladder II/WordLadderII.java
class Solution {
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        Set<String> dictionary = new HashSet<>(wordList);
        
        Set<String> visited = new HashSet<>();
        visited.add(beginWord);
        
        Queue<List<String>> queue = new LinkedList<>();
        List<String> beginList = new ArrayList<>();
        beginList.add(beginWord);
        queue.add(beginList);
        
        List<List<String>> results = new ArrayList<>();
        boolean found = false;
        
        while (!queue.isEmpty()) {
            int len = queue.size();
            Set<String> currentVisited = new HashSet<>();
            
            int i = len;
            while (i > 0) {
                List<String> currentSequence = queue.poll();
                String currentWord = currentSequence.get(currentSequence.size()-1);
                List<String> neighbors = findNeighbors(currentWord);
                for (String neighbor : neighbors) {
                    if (!visited.contains(neighbor) && dictionary.contains(neighbor)) {
                        List<String> newSequence = new ArrayList<>(currentSequence);
                        newSequence.add(neighbor);
                        if (neighbor.equals(endWord)) {
                            found = true;
                            results.add(newSequence);
                            continue;
                        }
                        currentVisited.add(neighbor);
                        queue.add(newSequence);
                    }
                }
                i--;
            }
            
            for (String word : currentVisited) {
                visited.add(word);
            }
            
            if (found) break;
        }
        
        return results;
    }
    
    private List<String> findNeighbors(String word) {
        char[] chars = word.toCharArray();
        List<String> neighbors = new ArrayList<>();
        
        for (int i = 0; i < word.length(); i++) {
            char current = word.charAt(i);
            for (int j = 0; j <= 25; j++) {
                char ch = (char) ((int) 'a' + j);
                chars[i] = ch;
                String neighbor = new String(chars);
                neighbors.add(neighbor);
            }
            chars[i] = current;
        }
        
        return neighbors;
    }
}

File: Problems/Algorithms/126. Word Ladder II/WordLadderII.kt
class Solution {
    fun findLadders(beginWord: String, endWord: String, wordList: List<String>): List<List<String>> {
        val dictionary: HashSet<String> = wordList.toHashSet()
        val visited: HashSet<String> = HashSet<String>()
        val queue = ArrayDeque<MutableList<String>>()
        val ans = mutableListOf<MutableList<String>>()
        var found = false
        
        queue.add(mutableListOf<String>(beginWord))
        visited.add(beginWord)
        
        while (!queue.isEmpty()) {
            val len = queue.size
            val currentVisited: HashSet<String> = HashSet<String>()
            
            var i = len
            while (i > 0) {
                val currentSequence = queue.removeFirst()
                val currentWord: String = currentSequence.last()
                val neighbors: List<String> = findNeighbors(currentWord)
                for (neighbor in neighbors) {
                    if (!visited.contains(neighbor) && dictionary.contains(neighbor)) {
                        val newSequence = currentSequence.toMutableList()
                        newSequence.add(neighbor)
                        if (neighbor.equals(endWord)) {
                            found = true
                            ans.add(newSequence)
                            continue
                        }
                        currentVisited.add(neighbor)
                        queue.add(newSequence)
                    }
                }
                i--
            }
            
            for (word in currentVisited) {
                visited.add(word)
            }
            
            if (found) break
        }
        
        return ans
    }
    
    private fun findNeighbors(string: String): List<String> {
        val chars = string.toCharArray()
        val neighbors = mutableListOf<String>()
        
        for (i in 0..string.length-1){
            val current = string.get(i)
            for (ch in 'a'..'z'){
                chars[i] = ch
                val neighbor: String = String(chars)
                neighbors.add(neighbor)
            }
            chars[i] = current
        }
        
        return neighbors
    }
}

